cmd/soroban-rpc/internal/db/db.go

```diff
@@ -353,3 +261,109 @@ func runMigrations(db *sql.DB, dialect string) error {
 	_, err := migrate.ExecMax(db, dialect, m, migrate.Up, 0)
 	return err
 }
+
+func (s *sqlDB) NewLedgerEntryReaderTx() (LedgerEntryReaderTx, error) {
+	opts := sql.TxOptions{
+		ReadOnly: true,
+	}
+	tx, err := s.db.BeginTxx(context.Background(), &opts)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ledgerReaderTx{
+		buffer: xdr.NewEncodingBuffer(),
+		tx:     tx}, nil
+}
+
+type ledgerReaderTx struct {
+	tx     *sqlx.Tx
+	buffer *xdr.EncodingBuffer
+}
+
+func (l ledgerReaderTx) GetLatestLedgerSequence() (uint32, error) {
+	sqlStr, args, err := sq.Select("value").From(metaTableName).Where(sq.Eq{"key": latestLedgerSequenceMetaKey}).ToSql()
+	if err != nil {
+		return 0, err
+	}
+	var results []string
+	if err = l.tx.Select(&results, sqlStr, args...); err != nil {
+		return 0, err
+	}
+	switch len(results) {
+	case 0:
+		return 0, ErrEmptyDB
+	case 1:
+	// expected length on an initialized DB
+	default:
+		panic(fmt.Errorf("multiple entries (%d) for key %q in table %q", len(results), latestLedgerSequenceMetaKey, metaTableName))
+	}
+	latestLedgerStr := results[0]
+	latestLedger, err := strconv.ParseUint(latestLedgerStr, 10, 32)
+	if err != nil {
+		return 0, err
+	}
+	return uint32(latestLedger), nil
+}
+
+func (l *ledgerReaderTx) GetLedgerEntry(key xdr.LedgerKey) (bool, xdr.LedgerEntry, error) {
```

Comment:
I'm not sure if I'm following this correctly; while executing a smart contract, the host would request a specific key.
we would then:
1. pull the value from the database ( and get a byte slice )
2. convert the byte slice into xdr.LedgerEntry
3. convert the xdr.LedgerEntry back to byte slice
4. encode the byteslice as base64
5. decode the base64 into bytes ( rust )
6. decode the bytes into xdr ( rust )

that's seems to be a lot of conversions. maybe we would have a simpler version of this method that just get the bytes and pass these, as-is, to the rust code ? i.e.
1. pull the value from the database ( and get a byte slice )
2. decode the bytes into xdr ( rust )

Reply:
You are right but this requires changing the database interface to return XDR-encoded ledger entries and passing buffers+lengths between Go and Rust.

I would like to only do that when/if needed due to performance reasons.

Reply:
For future reference, see https://stackoverflow.com/questions/64574831/how-to-return-struct-from-cgo

Reply:
Happy to create a ticket once we merge this to do this in the future.